Source/WebCore/html/shadow/TextControlInnerElements.cpp

```diff
@@ -240,14 +240,18 @@ Ref<SearchFieldResultsButtonElement> SearchFieldResultsButtonElement::create(Doc
 
 std::optional<Style::ElementStyle> SearchFieldResultsButtonElement::resolveCustomStyle(const Style::ResolutionContext& resolutionContext, const RenderStyle* shadowHostStyle)
 {
+    // The only purpose of this function is to cache the logical width, so it can be re-used during layout.
+
     RefPtr input = downcast<HTMLInputElement>(shadowHost());
-    if (input && input->maxResults() >= 0)
-        return std::nullopt;
+    bool isForcedAppearanceNone = !(input && input->maxResults() >= 0)
+        && shadowHostStyle && shadowHostStyle->effectiveAppearance() != SearchFieldPart;
+    SetForScope canAdjustStyleForAppearance(m_canAdjustStyleForAppearance, isForcedAppearanceNone);
 
-    if (shadowHostStyle && shadowHostStyle->effectiveAppearance() != SearchFieldPart) {
-        SetForScope canAdjustStyleForAppearance(m_canAdjustStyleForAppearance, false);
-        return resolveStyle(resolutionContext);
-    }
+    auto elementStyle = resolveStyle(resolutionContext);
+    if (elementStyle.renderStyle->display() == DisplayType::None)
+        m_cachedLogicalWidth = 0;
+    else
+        m_cachedLogicalWidth = elementStyle.renderStyle->logicalWidth().value();
 
     return std::nullopt;
```

Comment:
Why does this always return nullopt instead of the style we just computed (unlike the previous code)?

Why do we need to cache this in the first place instead of just querying from the RenderStyle of SearchFieldResultsButtonElement? Is there some case where the style is null but we still need these values?

Reply:
> Why does this always return nullopt instead of the style we just computed (unlike the previous code)?

Let me check if it's still needed.

> Why do we need to cache this in the first place instead of just querying from the RenderStyle of SearchFieldResultsButtonElement? Is there some case where the style is null but we still need these values?

The size depends on what happens in `SearchFieldResultsButtonElement::resolveCustomStyle`. Depending on `canAdjustStyleForAppearance`, style adjustments (width/height) can occur when calling `resolveStyle()`. The previous code querying `computedStyle()` doesn't go through `SearchFieldResultsButtonElement::resolveCustomStyle`, which ends up with the wrong dimensions.

Reply:
> The size depends on what happens in SearchFieldResultsButtonElement::resolveCustomStyle. Depending on canAdjustStyleForAppearance, style adjustments (width/height) can occur when calling resolveStyle(). The previous code querying computedStyle() doesn't go through SearchFieldResultsButtonElement::resolveCustomStyle, which ends up with the wrong dimensions.

Right but why can't the code just read element.renderer()->style() directly when it needs these values (and use 0 if there is no renderer in display:none case)?

Reply:
I guess we could, not sure why we were using `computedStyle()` in the first place (perhaps there's a reason, will need to look).